pkg/sr/serde.go

```diff
@@ -316,44 +323,108 @@ func (s *Serde) DecodeNew(b []byte) (any, error) {
 }
 
 func (s *Serde) decodeFind(b []byte) ([]byte, tserde, error) {
-	if len(b) < 5 || b[0] != 0 {
-		return nil, tserde{}, ErrBadHeader
+	h := s.Header()
+
+	id, b, err := h.DecodeID(b)
+	if err != nil {
+		return nil, tserde{}, err
 	}
-	id := binary.BigEndian.Uint32(b[1:5])
-	b = b[5:]
 
-	t := s.loadIDs()[int(id)]
+	t := s.loadIDs()[id]
 	if len(t.subindex) > 0 {
-		r := bReader{b}
-		br := io.ByteReader(&r)
-		l, err := binary.ReadVarint(br)
-		if l == 0 { // length 0 is a shortcut for length 1, index 0
-			t = t.subindex[0]
-		}
-		for err == nil && t.subindex != nil && l > 0 {
-			var idx int64
-			idx, err = binary.ReadVarint(br)
-			t = t.subindex[int(idx)]
-			l--
-		}
+		var index []int
+		index, b, err = h.DecodeIndex(b, t.subindexDepth)
 		if err != nil {
-			return nil, t, err
+			return nil, tserde{}, err
+		}
+		for _, idx := range index {
+			if t.subindex == nil {
+				return nil, tserde{}, ErrNotRegistered
+			}
+			t = t.subindex[idx]
 		}
-		b = r.b
 	}
 	if !t.exists {
-		return nil, t, ErrNotRegistered
+		return nil, tserde{}, ErrNotRegistered
 	}
 	return b, t, nil
 }
 
-type bReader struct{ b []byte }
+// SerdeHeader encodes and decodes a message header.
+type SerdeHeader interface {
+	AppendEncode(b []byte, id int, index []int) ([]byte, error)
+	DecodeID(in []byte) (id int, out []byte, err error)
+	DecodeIndex(in []byte, maxLength int) (index []int, out []byte, err error)
+}
+
+// ConfluentHeader is a SerdeHeader that produces the Confluent wire format. It
+// starts with 0, then big endian uint32 of the ID, then index (only protobuf),
+// then the encoded message.
+//
+// https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format
+type ConfluentHeader struct{}
```

Comment:
Can this be private? Not sure it needs to be public at the moment -- we can default internally to the private confluent header, but then the user can override.

Reply:
And, if we do that, this can be something like

```go
type tConfluentHeader struct{}

var confluentHeader = new(tConfluentHeader)
```

and then line 143 above can return `confluentHeader` the pointer variable (per the pointer methods comment below)

Reply:
I made the type private, but used different names to avoid the `t` prefix, hope that's fine:
```
type confluentHeader struct{}

var defaultSerdeHeader = new(confluentHeader)
```

Although if that's a convention you are using for private types (i.e. `tserde`, `tConfluentHeader` etc.) let me know.

Reply:
The t in this case means type, I do like it but I'm ok either way.